plugins/amqp_witness_plugin/amqp_witness_plugin.cpp

```diff
@@ -44,32 +36,23 @@ void amqp_witness_plugin::set_program_options(options_description& cli, options_
          ("witness-delete-unsent", boost::program_options::bool_switch(&my->delete_previous), "Delete unsent witness signature messages retained from previous connections");
 }
 
-void amqp_witness_plugin::plugin_initialize(const variables_map& options) {
-   try {
-      const auto& [pubkey, provider] = app().get_plugin<signature_provider_plugin>().signature_provider_for_specification(options["witness-signature-provider"].as<std::string>());
-      my->signature_provider = provider;
-   }
-   FC_LOG_AND_RETHROW()
-}
+void amqp_witness_plugin::plugin_initialize(const variables_map& options) {}
 
 void amqp_witness_plugin::plugin_startup() {
    const boost::filesystem::path witness_data_file_path = appbase::app().data_dir() / "amqp" / "witness.bin";
 
    if(boost::filesystem::exists(witness_data_file_path) && my->delete_previous)
       boost::filesystem::remove(witness_data_file_path);
 
-   my->rqueue = std::make_unique<reliable_amqp_publisher>(my->amqp_server, my->exchange, my->routing_key ? *my->routing_key : "", witness_data_file_path, "eosio.node.witness_v0");
-
-   app().get_plugin<chain_plugin>().chain().irreversible_block.connect([&](const chain::block_state_ptr& bsp) {
-      if(bsp->block->timestamp.to_time_point() < fc::time_point::now() - fc::seconds(my->staleness_limit))
-         return;
+   my->rqueue = std::make_unique<reliable_amqp_publisher>(my->amqp_server, my->exchange, my->routing_key, witness_data_file_path, "eosio.node.witness_v0");
 
-      my->rqueue->post_on_io_context([this, bsp]() {
-         try {
-            my->rqueue->publish_message(std::make_pair(bsp->header.action_mroot, my->signature_provider(bsp->header.action_mroot)));
-         }
-         FC_LOG_AND_DROP();
+   app().get_plugin<witness_plugin>().add_on_witness_sig([my=my](const chain::block_state_ptr& bsp, const chain::signature_type& sig) {
+      my->rqueue->post_on_io_context([my, bsp, sig]() {
```

Comment:
Whoops.  I missed the self-join that happens when `my` is the last reference to the `amqp_witness_plugin_impl`.  If you capture a raw pointer to `*rqueue`, I *think* it's safe, due to the destructor of `reliable_amqp_publisher`.